Pull request for project review

leonmezu1 opened the pull request:
# Project requirements:

 - [x]  Build a Rails application ensuring that:
    - The models, associations, controllers, and views are working as expected
    - Build successful sign-up, log-in and log-out methods within the app ensuring sessions behavior
    - Add additional restrictions based on detailed requirements

Further information is available at [The Odin Project Authentication](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)

Usagib committed:
staticpages controller generated added custom.scss

staticpages > home help login signup

Usagib committed:
AddModels 

User andPost model added

leonmezu1 committed:
Updated gemfile

Added needed gems for styling and authentication

leonmezu1 committed:
ran basic db migrate

Added user constraints and ran migrate to db schema

leonmezu1 committed:
Added association User <-> Post

The association User 1:N Post is working

leonmezu1 committed:
Basic routing for static done

The static  pages for the app skeleton are working as expected and the User authentication is working too

Usagib committed:
session controller added

leonmezu1 committed:
Added users controllers and views

In order to have a RESTful the users controllers and views were added that way the correct app behavior is guareanteed

leonmezu1 committed:
Fixed patch method for user form

users edit and new path are now working

leonmezu1 committed:
minor fixes

Fixed forms and refactored wrong paths

Usagib committed:
Session controller helper added

session login sucessfull

Usagib committed:
add remember_me_token column to user

db migration done

leonmezu1 committed:
Added authentication skeleton for session

replaced the bcrypt password method for a SHA1 digest hash

leonmezu1 committed:
Verified authentication method

Authentication method is working and the cookies were added

leonmezu1 committed:
Step 10 under test

setcurrent user and current user

leonmezu1 committed:
required fixes

Added some fixes to propperly kill the session

Usagib committed:
post model added

new post form added and linked to users

leonmezu1 committed:
Ran rubocop lint

Fixed project format ion order to pass the stickler tests

leonmezu1 committed:
Add gitpod config

this commit adds support for Gitpod.io, a free automated
dev environment that makes contributing and generally working on GitHub
projects much easier. It allows anyone to start a ready-to-code dev
environment for any branch, issue and pull request with a single click.

Usagib committed:
front-end bootstrap made

Usagib committed:
bootstrap layout fixes

leonmezu1 committed:
Fixed authoriztion on edit and update

Usagib committed:
error message layout changed

leonmezu1 committed:
Added gems for production

leonmezu1 committed:
Fixed problematic gem

Usagib committed:
Readme updated

maelfosso commented on config/routes.rb:
```diff
@@ -1,3 +1,18 @@
+  delete '/logout',  to: 'sessions#destroy'
+
+  resources :users
+  resources :posts
```
Update that file with the routes needed for that project and corresponding to the requirements

leonmezu1 replied:
## Routes definitions updated:
```
  root 'static_pages#home'
  get '/help', to: 'static_pages#help'

  get '/new_post', to: 'posts#new'
  post '/new_post', to: 'post#create'
  get '/posts', to: 'posts#index'

  get    '/login',   to: 'sessions#new'
  post   '/login',   to: 'sessions#create'
  delete '/logout',  to: 'sessions#destroy'

  resources :users, only: %i[show index]
  resources :posts, only: %i[new create index]
```

maelfosso commented on app/models/user.rb:
```diff
@@ -0,0 +1,59 @@
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
instead of `remember_digest` use `remember_token`

leonmezu1 replied:
## Updated attribute

```
  def create_remember_token
    rem_token = Digest::SHA1.hexdigest(SecureRandom.urlsafe_base64)
    self.remember_token = rem_token
  end 
```

maelfosso commented on app/models/user.rb:
```diff
@@ -0,0 +1,59 @@
+  # def rtest(string)
+  # self.token = string
+  # update_attribute(:remember_digest, User.digest(token))
+  # end
```
Remove the commented code and use `git` to access to the history

leonmezu1 replied:
## Cleared

maelfosso commented on app/models/user.rb:
```diff
@@ -0,0 +1,59 @@
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
```
For that project, you can merge them into one method so that it return  the digest of a token directly.

leonmezu1 replied:
## Refactored

```
  def create_remember_token
    rem_token = Digest::SHA1.hexdigest(SecureRandom.urlsafe_base64)
    self.remember_token = rem_token
  end 
```

maelfosso commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    user = find_user
+    if user&.authenticate(params[:session][:password])
+      log_in user
+      remember user
```
`log_in user` is not necessary. `remember user` is sufficient.

leonmezu1 replied:
## Deleted unnecessary method

maelfosso commented on app/helpers/session_helper.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def current_user
+    @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]
+  end
```
Use `cookies.permanent` here instead

leonmezu1 replied:
## Replaced for cookies method

```
  def current_user
    @current_user ||= User.find_by(id: cookies.permanent.signed[:user_id]) if cookies.permanent.signed[:user_id]
  end
```

maelfosso commented on app/helpers/session_helper.rb:
```diff
@@ -0,0 +1,49 @@
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
```
At the end, set the current user to the user sent in parameter.

leonmezu1 replied:
## User setter added

maelfosso commented on app/helpers/session_helper.rb:
```diff
@@ -0,0 +1,49 @@
+    else
+      @current_user = nil
+    end
+  end
```
Not really needed. You can use one method up .

leonmezu1 replied:
## Deleted unnecessary method

maelfosso commented on app/helpers/session_helper.rb:
```diff
@@ -0,0 +1,49 @@
+    session.delete(:user_id)
+    @current_user = nil
+  end
+end
```
NO session, set cookies.permanent to nil

leonmezu1 replied:
## Refactored

```
  def log_out
    cookies.permanent.signed[:user_id] = nil
    cookies.permanent[:remember_token] = nil
    @current_user = nil
  end
``

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello,
Your project is awesome, the design is original, amazing.
However, there are some requirements that need to be met.

1. Update your README. There is no user, no sign up link. So we expect to see a description in your README about how to run the app.

2. From that [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png), it is indicated that the application routes which is different from what you've defined

3. No need for `remember_digest`. The `remember_token` you defined is sufficient.

4. Please, **no use of `session`** in all the project. **Forget about it completely.**
> The remember_me token the project describes consists of storing a hash in a cookie and using it to determine in a user is logged in. Do not confuse it with the Remember me [ ] checkbox that many websites show. That is a more complex solution, which is not required here

So, for authentication, just set `cookies.permanent[:remember_token]` with the **user's remember_token**. If it's defined there is an user connected and you can use `current_user` to know who is or define him. If it's nil, there is no user. Take a look in that [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png).

5. Read the comments below.

Kindly make the changes and submit a new code review request.
Thanks

Mael FOSSO

Usagib committed:
session changed for cookie permanent login

leonmezu1 committed:
Updated routes and schema

Usagib committed:
Merge remote-tracking branch 'origin/post-auth' into post-auth

Usagib committed:
updated remember_token atrribute

leonmezu1 committed:
Final fixes done

The help view is improved and the disgest method is now refactored the application is running and ready to the action!

leonmezu1 commented:
Required changes submitted. :v:

leonmezu1 committed:
Update README.md

Some extra running instructions

leonmezu1 committed:
Update README.md

Fixed small typos

wathigo reviewed (changes requested):
### Changes Required
Hello Devs :wave: 
The design looks amazing :sparkles: 
All the requested changes have been met.
Let's address one issue.
**[REQUIRED]**
* User can sign up ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_up_requirement.png) from Pathwright).

**After implementing the change, submit another request for review.**

**Good luck :four_leaf_clover: 
Simon Wathigo**

leonmezu1 commented:
## Subsection 6

Subsection 6 specifies:

  - Create a couple of users to populate your app with. **We won’t be building a sign up form**, so you’ll need to create new users via the command line. Your #before_create should now properly give each newly created user a special token.

leonmezu1 committed:
Sign up feature is  now ready to go!

leonmezu1 commented:
## Sign up

  - Feature reassembled

wathigo reviewed (approved):
### Approved :heavy_check_mark: 
Great work implementing the requested changes :clap: 

**You may proceed to the next project.**

**Good luck :four_leaf_clover: 
Simon Wathigo**
